Common problems faced by PHP developers in consuming an AMQP message (#22)
facile-it/facile-it.github.io: content/blog/eng/common-problems-faced-by-php-developers-in-consuming-an-ampq-message.md

---
authors: ["emulator000"]
date: "2018-11-15"
draft: true
share: true
categories: [English, RabbitMQ, AMQP, Consumer, PHP, Rust, Go]
title: "Common problems faced by PHP developers in consuming an AMQP message"
type: "post"
languageCode: "en-EN"
toc: true
---

When we use RabbitMQ and our project is in PHP, we have to run a Consumer and we could encounter some common and still unresolved problems.

> In order to implement RabbitMQ in the project, there are different ready to use packages, especially when we use a framework like Symfony or Laravel. Usually these packages provides the full needed integration for the AMQP protocol so, we can easily configure it and create a Producer, a Consumer and then, we have to create a little script that will supervisor the PHP process in order to make sure that all our consumers are running. 

Reviewer: ```suggestion
In order to implement RabbitMQ in the project, there are different ready to use packages, especially when we use a framework like Symfony or Laravel. Usually these packages provide the full integration needed for the AMQP protocol, so we can easily configure it and create a Producer and a Consumer; then, we have to create a little script that will supervise the PHP process in order to make sure that all our consumers are running. 
```

# Consumer

> In order to consume a message, we have to run a demon that will listen for new messages and then, process each message executing some logic; this is a normal flow in case of blocking requests, we don't want to block the user request nor loose it.

Reviewer: ```suggestion
In order to consume a message, we have to run a demon that will listen for new messages, then process each message executing some logic: this is a normal flow in case of blocking requests, and we don't want to block the user request nor loose it.
```

> We want also to keep track of all requests and if one fails, we could retry to process again or just ignore it if some blocking error occurs.

Reviewer: ```suggestion
We want also to keep track of all requests and if one fails, we could retry to process it again or just ignore it if some blocking error occurs.
```

## The memory problem

> A normal PHP script serves a single request, then, the script "dies" naturally as it ended the execution of provided instructions and frees the allocated memory. In a PHP consumer, we have to run a long running PHP process so we have to face the memory problem, yes, because as PHP is not properly designed to achieve this scope, this can cause troubles for sure if we don't pay attention to the script's code.

Reviewer: ```suggestion
A normal PHP script serves a single request, then the script _dies_ naturally as the execution of provided instructions is completed, and the allocated memory is freed. In a PHP consumer we have to execute a long running PHP process: this will cause us to face a memory problem, because PHP is not properly designed to achieve this goal, and this will cause trouble for sure if we don't pay attention to the script code.
```

> Normally, you don't notice any memory issue due to fact that the script runs for few seconds, also, in a long running PHP proccess, if you try a simple script and make it run for few minutes, maybe they will not use any additional memory at all but if you use some framework, an ORM or you are in a complex application with many dependencies, the application will eventually allocate some memory (eg. for caching results) and you can notice that the PHP process will increase their memory hour by hour due to some memory leaks.

Reviewer: ```suggestion
Normally, you don't notice any memory issue due to fact that the script runs for few seconds; also, in a long running PHP process, if you try and run a simple script for few minutes it could not use any additional memory at all, but if you use some framework, an ORM or you are in a complex application with many dependencies, the application will eventually allocate some memory (e.g. for caching results), and you could even notice a progressive increase of the memory used by the PHP process due to some memory leaks.
```

> That's not a problem at all in a normal case, the script ends and the memory is freed normally, this is something that is not so obvious in a long running process like how it would be a PHP consumer. They will break, leak and in some cases, corrupts the memory and this will make the process crash or more drastically, they will fail to run some instruction as the opcode cache gets changed while the process is running.

Reviewer: ```suggestion
That's not a problem at all in a normal case, where the script ends and the memory is freed normally, but the case is not that obvious for a long running process like the one required by a PHP consumer. It will break, leak and, in some cases, corrupt the memory, and this will make the process crash or, more drastically, fail to run some instructions, as the opcode cache gets changed while the process is running.
```

> If you are interested in more accurate details and information, I suggest [this article](http://notjam.es/blog/2014/06/18/the-problems-with-long-running-php.html), is pretty old but explains this issue in a very well and accurate way.

Reviewer: ```suggestion
If you are interested in more accurate details and information, I suggest reading [this article](http://notjam.es/blog/2014/06/18/the-problems-with-long-running-php.html): it's pretty old, but explains this issue very well.
```

## Multiple consumers

> If you are running a PHP consumer, is very hard to process multiple messages at time because PHP runs in a single thread and also, a proper async/await interface is not well supported or mature. One more time, PHP is not well designed for this scope and this could be a limit in some circumstances, scaling is very hard.

Reviewer: ```suggestion
For a PHP consumer it's very hard to process multiple messages at the same time, because PHP runs in a single thread, and a proper async/await interface is not well supported or mature. Again, PHP is not designed for this scope and this could be a limit in some circumstances: scaling is really hard.
```

> Running multiple PHP consumers is not a good idea as well, supposing a framework based application, this could use an huge amount of memory and could reach easily GBs of RAM used. If in the "waiting" time there are no messages, this is a very waste of memory that instead, could be used for some other resources or requests. 

Reviewer: ```suggestion
Running multiple PHP consumers is not a good idea as well: given a framework-based application, it could use a huge amounts of RAM, easily GBs. If during the "waiting" time there are no messages, it's going to be a serious waste of memory that could be used instead for some other resources or requests. 
```

## Updating the codebase

> Another common problem when we have PHP consumers is when we have to update the codebase: if we deploy something that change the code and some service used by the consumer, we can easily break the integrity of the entire long running process execution.

Reviewer: ```suggestion
Another common problem with PHP consumers is when we need to update the codebase: if we deploy something that changes the code and some service used by the consumer, we can easily break the integrity of the entire long running process execution.
```

> We have to close all the consumers and sometimes we have to force kill processes, because, they hangs due to some memory corruption, so, we have to start them again and obviously, we have to provide a script that will do this automatically for us during the deploy or in the pipeline.

Reviewer: ```suggestion
We have to close all the consumers and sometimes we must force kill processes: they could hang due to memory corruption, so we have to start them again with a script (obviously) that will trigger automatically during the deploy or within the pipeline.
```
